src/app/services/country.service.ts

```diff
@@ -1,54 +1,56 @@
 import { Injectable } from '@angular/core';
 import { HttpClient } from '@angular/common/http';
-import { Observable, map, of, shareReplay } from 'rxjs';
+import { Observable, of, map, shareReplay, catchError } from 'rxjs';
 import { Country, CountryDetail } from '../models/country.model';
 
 @Injectable({
   providedIn: 'root',
 })
 export class CountryService {
   private apiUrl = 'https://restcountries.com/v3.1/all';
-  private countries$: Observable<Country[]>;
+  private searchByNameApiUrl = 'https://restcountries.com/v3.1/name'; // API for search by name
+  private searchByCapitalApiUrl = 'https://restcountries.com/v3.1/capital'; // API for search by capital
 
-  constructor(private http: HttpClient) {
-    this.countries$ = this.http
-      .get<Country[]>(this.apiUrl)
-      .pipe(shareReplay(1));
+  constructor(private http: HttpClient) {}
+  // Get all countries from cached API call
+  getAllCountries() {
+    return this.http.get<any>(this.apiUrl);
   }
 
-  // get country data
-  getAllCountries(): Observable<Country[]> {
-    return this.countries$;
-  } 
-
-  // get perticular data when open dialog
+  // Get specific country by name from the cached data
   getCountryByName(name: string): Observable<CountryDetail | any> {
-    return this.countries$.pipe(
-      map(
-        (countries) =>
-          countries.find((country) => country.name.common === name) || null
-      )
-    );
-  }
-
-  // get search data
-  searchCountries(searchTerm: string): Observable<Country[]> {
-    return this.countries$.pipe(
-      map((countries) =>
-        countries.filter((country) =>
-          this.matchesSearchTerm(country, searchTerm)
+    return this.http
+      .get<any>(this.apiUrl)
+      .pipe(
+        map(
+          (countries) =>
+            countries.find((country: any) => country.name.common === name) ||
```

Comment:
Why are we still finding the country here from the list? Why not use the find api like -https://restcountries.com/v3.1/name/eesti